fix ":menu Item.SubItem", fix menu_get("foo")

justinmk opened the pull request:
The following steps should print the `File.Save` submenu:

```
nvim -u NORC
:source $VIMRUNTIME/menu.vim
:menu File.Save
```

This was regressed in dc685387a3d6. Blocks https://github.com/neovim/neovim/pull/8173 .

`menu_get()` also was missing some results for some cases (covered by existing tests, see below).

justinmk committed:
fix ":menu Item.SubItem"

:menu should print sub-menu contents. E.g. this should print the
"File.Save" submenu:

    nvim -u NORC
    :source $VIMRUNTIME/menu.vim
    :menu File.Save

Regressed in dc685387a3d6
Blocks #8173

menu_get() also was missing some results for some cases.

justinmk commented on test/functional/ex_cmds/menu_spec.lua:
```diff
@@ -363,14 +363,24 @@ describe('menu_get', function()
+            hidden = 0
+          },
+          {
+          },
```
The updated result looks correct to me because "Test" should match the `Test` and `Test.Test` submenu.

But the empty item at the end obviously needs to be fixed.

justinmk commented:
Besides the note above, there is one failing test which I haven't fixed. @teto can you look at this?

```
|| [  FAILED  ] test/functional/ex_cmds/menu_spec.lua @ 309: menu_get matching path, default modes
|| test/functional/ex_cmds/menu_spec.lua:328: Expected objects to be the same.
|| Passed in:
|| (table) {
||  *[1] = {
||     [hidden] = 0
||     [name] = 'Export'
||     [priority] = 500
||     [submenus] = {
||       [1] = { ... more } } }
||   [2] = {
||     [hidden] = 0
||     [name] = 'Edit'
||     [priority] = 500
||     [submenus] = {
||       [1] = { ... more } } }
||   [3] = {
||     [hidden] = 1
||     [name] = ']Export'
||     [priority] = 500 
||     [submenus] = {
||       [1] = { ... more } } } }
|| Expected:  
|| (table) {  
||  *[1] = {  
||     [hidden] = 0
||     [mappings] = {
||       [n] = { ... more } }
||     [name] = 'Script'
||     [priority] = 500
||    *[tooltip] = 'This is the tooltip' } }
||-
|| stack traceback:
|| >    test/functional/ex_cmds/menu_spec.lua:328: in function <test/functional/ex_cmds/menu_spec.lua:309>
```

teto commented:
thanks for the detailed steps, I will have  a look a it

janlazo commented:
Is this blocked?

justinmk commented:
@janlazo I will take another look at it today. I might just  regress `menu_get()`slightly (it will have extra items) so that `:menu` works correctly. Then hopefully @teto can fix it later.

teto commented:
@janlazo sorry for blocking your vim porting effort. I will do my best to fix it this week.

janlazo commented:
@teto It's okay. I have other patches to port in the meantime. @ckelsel PR is blocked so I asked. I'd like to resolve the Vim patches up to 8.0.0693 but I'll likely not make it in time for 0.3.1.

janlazo commented:
Any updates?

justinmk commented:
@teto do you have time to look at this?

janlazo commented:
What's the use case for `$VIMRUNTIME/menu.vim`? I don't recall using this on gvim (if it relies on `guioptions` and I don't source it on my vimrc.

justinmk committed:
menu_get(): Do not include empty items

Caused by a typo: `dict` instead of `dic`. Renamed variable to `d` to
make it less similar.

justinmk commented:
Fixed the bugs mentioned above.

> What's the use case for $VIMRUNTIME/menu.vim?

@janlazo 	It should be sourced by default, see `:help menu.vim`. However `nvim -u NORC` does not source it. Possibly we stopped doing this by default because it isn't relevant to move Nvim UIs. 

Also have never heard a user request it. `:helpgrep` and tab-completion at the `:` prompt is usually easier ...

justinmk committed:
menu_get(): fix query behavior

- Return the menu properties, not only its children.
- If the {path} param is given, return only the first node. The "next"
  nodes in the linked-list are irrelevant.
